### PR TITLE
Advanced feature functions

### PR DESCRIPTION
**Pull request by halfak**
Extends comparison operators to include {+, -, /, *} as well as {==, !=, <, >, <=, >=}.  Also adds modifiers for min() and max().

This includes a full test suite and a slightly more elegant strategy for implementing the operators.

### CODE REVIEW

File: revscoring/features/tests/test_feature.py [hunk 1]
```diff
@@ -7,17 +7,27 @@
...
 
 def identity_process(value): return value
 int_identity = Feature("int_identity", identity_process,
                                        returns=int, depends_on=["value"])
 
+def check_feature(feature, expected):
+
+    eq_(feature.returns, type(expected))
+
+    eq_(hash(pickle.loads(pickle.dumps(feature))), hash(feature))
```
Comment (he7d3r) [inline revscoring/features/tests/test_feature.py]: Won't this use of `hash` have problems such as http://stackoverflow.com/a/793835/2062663 ?

File: revscoring/features/feature.py [hunk 2]
```diff
@@ -100,46 +132,94 @@ class mul(BinaryOperator):
     
     CHAR = "*"
     
-    def operate(self, left, right): return left * right
+    def operate(self, left, right):
+        print(str(left) + " * " + str(right))
```
Comment (he7d3r) [inline revscoring/features/feature.py]: print

File: revscoring/features/feature.py [hunk 3]
```diff
@@ -100,46 +132,94 @@ class mul(BinaryOperator):
...
+                         depends_on=dependencies)
+    
+    def _process(self, *feature_values): return float(math_max(*feature_values))
 
-class sub(Modifier):
+class min(Modifier):
+    
+    def __init__(self, *args):
+        dependencies = [Feature.or_constant(arg) for arg in args]
+        returns = float # Hardcoded even though max can return strings, it
```
Comment (he7d3r) [inline revscoring/features/feature.py]: We have `returns=float` on line 212, so this one seems unused

File: revscoring/features/feature.py [hunk 4]
```diff
@@ -100,46 +132,94 @@ class mul(BinaryOperator):
...
+                        # shouldn't ever do that
+        
+        name = "min({0})".format(", ".join(f.name for f in dependencies))
+        super().__init__(name, self._process, returns=float,
```
Comment (he7d3r) [inline revscoring/features/feature.py]: Was this supposed to be `returns=returns`?

File: revscoring/features/feature.py [hunk 5]
```diff
@@ -100,46 +132,94 @@ class mul(BinaryOperator):
...
+    CHAR = "!="
+    
+    def operate(self, left, right): return left != right
 
-def _process(self, feature_value): return feature_value + self.summand
+class max(Modifier):
+    
+    def __init__(self, *args):
+        dependencies = [Feature.or_constant(arg) for arg in args]
+        returns = float # Hardcoded even though max can return strings, it
```
Comment (he7d3r) [inline revscoring/features/feature.py]: We have `returns=float` on line 199, so this one seems unused

File: revscoring/features/feature.py [hunk 6]
```diff
@@ -100,46 +132,94 @@ class mul(BinaryOperator):
...
+                        # shouldn't ever do that
+        
+        name = "max({0})".format(", ".join(f.name for f in dependencies))
+        super().__init__(name, self._process, returns=float,
```
Comment (he7d3r) [inline revscoring/features/feature.py]: Was this supposed to be `returns=returns`?

File: revscoring/features/tests/test_feature.py (hunk 1 above)
Comment (halfak) [inline revscoring/features/tests/test_feature.py]: This test is designed to make sure that it does not have that problem.

File: revscoring/features/feature.py (hunk 3 above)
Comment (halfak) [inline revscoring/features/feature.py]: Fixed

File: revscoring/features/feature.py (hunk 5 above)
Comment (halfak) [inline revscoring/features/feature.py]: Fixed

**Pull request merged**